### DISCUSSION TITLE
Why is the $spacer variable defined using REM?

### DISCUSSION THREAD

**Discussion by tannermann** [twbs/bootstrap#35820] [General] [answered]
### Discussed in https://github.com/twbs/bootstrap/discussions/35461

<div type='discussions-op-text'>

<sup>Originally posted by **thenomadicmann** December  2, 2021</sup>
What is the point of defining the `$spacer` variable using REM units instead of pixels? If a user changes their browser's default font-size settings to be able to read text easier, they've also now increased margins and paddings which may crowd a design and possibly break a page layout. Why not reserve using REM for font-size? On a related note, in BS4, the grid gutter width, `$grid-gutter-width`, was set using pixels. In BS5 it switched to REM as well which causes the same unexpected behavior described above.</div>

**Comment by ffoodd** [comment]
`rem` is a relative unit, meaning everything sized with it will resize accordingly to the root font size. So for anyone changing their base font size, layout and content will resize and *preserve their proportion altogether*. 

When used correctly, zooming in and out will even trigger breakpoints, thus switch responsive layout.

Would you be able to provide a reduced test case demonstrating an issue?

**Comment by tannermann** [comment]
I'll have a harder time demonstrating a broken design I think, but here is an example of the undesirable consequences of REM for spacing:

Here is a page section with a medium font-size set in browser settings:
<img width="1359" alt="Screen Shot 2021-12-13 at 2 41 40 PM" src="https://user-images.githubusercontent.com/17735264/145885930-d8839d19-2527-4799-bbe1-dd1cb0caf306.png">

Here is the section after increasing the default browser font-size:
<img width="1282" alt="Screen Shot 2021-12-13 at 2 42 00 PM" src="https://user-images.githubusercontent.com/17735264/145885993-97a0359d-ea99-4f30-84d8-8bba346248ed.png">

When I adjust font-size, I expect the font-size of elements on a page to increase. What I don't expect or desire is for my visuals (images, etc) to be more spaced out. The padding on the left of the text also increases unnecessarily. The increased space in the grid and other padding means less space for my visuals and text.

**Comment by ffoodd** [comment]
How are you sizing your cards, in this example? Would you mind creating a CodePen or something that'd allow to have a closer look?

**Comment by tannermann** [comment]
In the screenshot below, you can see the padding utility `p-*` classes rely on the `$spacer` variable. I hope this helps. This is a quick example. We have many custom components and content sections throughout our site that use similar spacing methods. Other jobs I've had teams have done pretty much the same thing.

![Screen Shot 2021-12-13 at 3 20 08 PM](https://user-images.githubusercontent.com/17735264/145890582-53c32cc6-17ed-417d-8845-88cc1982ae7a.png)

**Comment by tannermann** [comment]
Regarding your point of using rem for element spacing to "preserve their proportion altogether", this seems to make too many assumptions and seems to do more than what a user might expect. I.e increasing the browser font-size should keep the existing layout and only adjust font-sizes.

**Comment by WinterSilence** [comment]
https://github.com/twbs/bootstrap/blob/28f7c94475b9456897ef9a4b60bc0f45d4db3388/scss/_variables.scss#L543

**Comment by JSieffert** [comment]
@thenomadicmann 

Without seeing everything at play here I can only guess what's going on but it appears these two elements, class.card-img and class.card-img-overlay, are positioned. The card-img is essentially the background and the card-img-overlay is bottom-left aligned. 

If I'm right, the bottom-left alignment is your issue. Align it from the top ensuring you leave enough space for the image to fully display and not cause the elements to overlap. 

e.g. card-img-overlay { top: 15rem; left: 0;}

**Comment by tannermann** [comment]
I wanted to follow this up with the fact that in BS4 the column gutter width is set with pixels (`$grid-gutter-width: 30px !default;`). In BS5, it is set using REM (`$grid-gutter-width: 1.5rem !default;`).

**Comment by asyncdesign** [comment]
Are you setting $font-size-root instead of $font-size-base somewhere ?

**Comment by mdo** [comment, marked as answer]
We long ago made the decision to move nearly all our units to `rem`s. The grid was a leftover as an assumption that, similar to the media queries not rendering properly with `rem`s, that we needed to keep some pixel values around. Spacers affect the `margin` and `padding` of type elements like headings and paragraphs, and provide more consistent scaling, so it makes sense they should be set in `rem`s like everything else.

**Edit:** Also, I turned this into another discussion as this wasn't an issue (bug or feature request).

**Reply by tannermann** [reply]
While I don't agree with having the margin/padding utilities use relative units, I understand and can see pros/cons.

Changing the grid gutters to rely on `rem` in BS5 seems more controversial. Why not carry pixels forward on `$grid-gutter-width` on BS5 as well instead of now defining it as rem? https://getbootstrap.com/docs/5.1/layout/grid/#customizing-the-grid

**Reply by mdo** [reply]
Because of the spacing classes being relative. If we put the grid in pixels, we basically had no way of using margin/padding utilities responsively to modify the grid. That was the main reason—`30px` of grid gutter in v4 couldn't be restored exactly to its original value if a utility was used on it. In addition, this keeps everything (grid and typography) using the same sizing scale.
